### PR TITLE
[SPARK-47015][SQL] Disable partitioning on collated columns

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?

Disable hive style partitioning on columns that are non default collated

### Why are the changes needed?

With current implementation partitioning on columns that have either accent or case insensitive collation would lead to incorrect results.

### Does this PR introduce _any_ user-facing change?

Only compared to the latest master version

### How was this patch tested?

With new UTs

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by MaxGekk — COMMENTED** [review]
@stefankandic Tags have some meaning actually. Where did you get `[Collation]` from?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 1]
```diff
@@ -174,4 +174,36 @@ class CollationSuite extends QueryTest with SharedSparkSession {
...
+      createTable("id")
+      createTable("c2")
+      createTable("id", "c2")
+
+      Seq(Seq("c1"), Seq("c1", "id"), Seq("c1", "c2")).foreach { partitionColumns =>
+        checkError(
+          exception = intercept[AnalysisException] {
+            createTable(partitionColumns: _*)
+          },
+          errorClass = "_LEGACY_ERROR_TEMP_1153",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:202]: As you are here, could you assign proper name for the error class and improve error message format (if it is needed).

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
how about bucket columns? We generate the bucket id from the string value and assume all the semantically-same string values should generate the same bucket id, which isn't true for string with collation.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 1 above)
Comment (stefankandic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: sure

**Comment by stefankandic** [comment]
> how about bucket columns? We generate the bucket id from the string value and assume all the semantically-same string values should generate the same bucket id, which isn't true for string with collation.

@mihailom-db this seems like a fairly straightforward task so can you take a look at it when you have the time? Should be not much different than this change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 2]
```diff
@@ -574,6 +573,14 @@ object PartitioningUtils extends SQLConfHelper {
     }
   }
 
+  /**
+   * Checks whether a given data type can be used as a partition column.
+   */
+  def canPartitionOn(dateType: DataType): Boolean = dateType match {
+    case st: StringType => st.isDefaultCollation
```
Comment (dbatomic) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:580]: Hey, I know that Spark codebase is not big on comments, but maybe we should add a comment why partitioning is disabled + jira ticket to enable the partitioning :) ?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 3]
```diff
@@ -174,4 +174,36 @@ class CollationSuite extends QueryTest with SharedSparkSession {
...
+        sql(
+          s"""
+             |CREATE TABLE $tableName
+             |(id INT, c1 STRING COLLATE 'UNICODE', c2 string)
+             |USING parquet
+             |PARTITIONED BY (${partitionColumns.mkString(",")})
+             |""".stripMargin)
+      }
+
+      // should work fine on non collated columns
```
Comment (dbatomic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:192]: Nit - on default collated columns :)

**Review by dbatomic — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala (hunk 2 above)
Comment (stefankandic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:582]: added the comment but not sure what you meant regarding the jira ticket to enable the partitioning - there is no plan to do this AFAIK and also I don't see how it would even be possible unless we revamp the entire partitioning logic - ie. not write the actual value to the file system

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -388,6 +388,11 @@
     ],
     "sqlState" : "42846"
   },
+  "CANNOT_USE_DATA_TYPE_FOR_PARTITION_COLUMN" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:391]: INVALID_PARTITION_COLUMN_DATA_TYPE?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 5]
```diff
@@ -574,6 +573,16 @@ object PartitioningUtils extends SQLConfHelper {
     }
   }
 
+  /**
+   * Checks whether a given data type can be used as a partition column.
+   */
+  def canPartitionOn(dateType: DataType): Boolean = dateType match {
+    // non default collated strings should not be used as partition columns
+    // as it could lead to incorrect results
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:581]: let's be more specific. "... as we cannot implement string collation semantic with directory names."

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 6]
```diff
@@ -174,4 +174,36 @@ class CollationSuite extends QueryTest with SharedSparkSession {
...
   }
+
+
+  test("disable partition on collated string column") {
+    def createTable(partitionColumns: String*): Unit = {
+      val tableName = "test_partition_tbl"
+      withTable(tableName) {
+        sql(
+          s"""
+             |CREATE TABLE $tableName
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:185]: can we also test v2 tables? See `DataSourceV2SQLSuite` about how we create and test v2 tables using v2 catalogs.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 6 above)
Comment (stefankandic) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:299]: it seems that we don't even check if partitioning is allowed or not for v2 table creation ([code](https://github.com/apache/spark/blob/a939a7d0fd9c6b23c879cbee05275c6fbc939e38/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L251)), is this a bug?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 7]
```diff
@@ -574,6 +573,16 @@ object PartitioningUtils extends SQLConfHelper {
...
   }
 
+  /**
+   * Checks whether a given data type can be used as a partition column.
+   */
+  def canPartitionOn(dateType: DataType): Boolean = dateType match {
+    // non default collated strings should not be used as partition columns
+    // as we cannot implement string collation semantic with directory names
+    case st: StringType => st.isDefaultCollation
+    case other => other.isInstanceOf[AtomicType]
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala:583]: we missed the `if !a.isInstanceOf[VariantType]` part

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:299]: nvm, v2 table does not use directory names to implement partitioning. We don't need to test it.

**Review by cloud-fan — APPROVED** [review]
LGTM except for one mistake

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala (hunk 7 above)
Comment (stefankandic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: thanks!

File: common/utils/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -2304,6 +2304,12 @@
     },
     "sqlState" : "22023"
   },
+  "INVALID_PARTITION_COLUMN_DATA_TYPE" : {
+    "message" : [
+      "Cannot use <type> for partition column."
+    ],
+    "sqlState" : "0A000"
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:2311]: cc @srielau for the sqlstate.

File: common/utils/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -5142,7 +5143,7 @@
   },
   "_LEGACY_ERROR_TEMP_1183" : {
     "message" : [
-      "Cannot use interval type in the table schema."
+      "Cannot use \"INTERVAL\" type in the table schema."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:5146]: BTW, it is unrelated changes, it seems.

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:5146]: If it is not related, please, revert this and the modified tests.

Comment (stefankandic) [reply common/utils/src/main/resources/error/error-classes.json:5146]: it isn't, there is a test `SQLInsertTestSuite.scala` checking for both cases where interval is used in a schema or in a partition column, so when I updated the error class for partitioning to use sql type it broke the test for checking for interval type in the schema

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:5146]: ok. I see.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @stefankandic and @dbatomic @cloud-fan for review.

File: common/utils/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -2304,6 +2304,12 @@
     },
     "sqlState" : "22023"
   },
+  "INVALID_PARTITION_COLUMN_DATA_TYPE" : {
+    "message" : [
+      "Cannot use <type> for partition column."
+    ],
+    "sqlState" : "42601"
```
Comment (srielau) [inline common/utils/src/main/resources/error/error-classes.json:2311]: Actually the 0A000 was more correct.
0A000 means a feature is not supported. It implies that in some future it may be and the user hit a limit rather than just writing something inherently wrong.

Comment (stefankandic) [reply common/utils/src/main/resources/error/error-classes.json:2311]: thanks for the response! will fix it in a separate PR since this one is already closed

Comment (dbatomic) [reply common/utils/src/main/resources/error/error-classes.json:2311]: Also let's stay consistent with error messages. E.g. in [string functions PR ](https://github.com/apache/spark/pull/45216) @uros-db is also using `0A000` for as `sqlState` so let us align there.
